torch/csrc/jit/codegen/cuda/scheduler/mma_utils.cpp

```diff
@@ -259,39 +166,165 @@ void WarpMmaSwizzler::setWarpMapped(TensorView* tv, int number_of_dims) {
+    bool include_this_id = false;
+    bool is_broadcast_in_a = a_domain[id_idx]->isBroadcast();
+    bool is_broadcast_in_b = b_domain[id_idx]->isBroadcast();
+    bool is_reduction_id = accumulator_domain[id_idx]->isReduction();
+
+    switch (dimension) {
+      case MmaDimension::K:
+        // K dimension is the dimension that is concrete in
+        //  operands, and is reduced by mma. This complies with
+        //  tensor contraction definition.
+        include_this_id =
+            !is_broadcast_in_a && !is_broadcast_in_b && is_reduction_id;
+        break;
+      // M and N dimension below are defined as the iterdomains
+      //  that are not reduced by mma, and are concretized in this stage.
+      case MmaDimension::M:
+        include_this_id =
+            !is_broadcast_in_a && is_broadcast_in_b && !is_reduction_id;
+        break;
+      case MmaDimension::N:
+        include_this_id =
+            is_broadcast_in_a && !is_broadcast_in_b && !is_reduction_id;
+        break;
+
+      default:
+        TORCH_INTERNAL_ASSERT(false, "unreachable");
+    }
+
+    if (include_this_id) {
+      result.push_back(accumulator_domain.at(id_idx));
+    }
+  }
+
+  return result;
+}
+
+// [MMA dimension matching]
+// Returns all the axes that correspond to the given mma dimension. This is the
+// first relaxation step
+//  on the mma check. Mma operations concerns 3 dimensions, namely, the M, N,
+//  and K dimension, more details see [Operand Layout Convention] in mma_type.h.
+//  The current implementation, for best effort safety, supports the patterns
+//  where the root axes can be classified into one of the 3 dimension types.
+//  This is a helpful initial step into defining tensor contraction
+//  optimizations.
+//
+// A concrete example:
+//  T0 [I0, I1, I2, I3, I4, I5] = mma(T1[I01, B11, B21, I31, I41, B51], T2[B02,
+//  I12, B22, I32, I42, I52], {3};
+// In this case some example querries:
+//  K dimension of T0 = {I3}
+//  M dimension of T1 = {I01}
+//  N dimension of T2 = {I52}
+//  etc.
+std::vector<IterDomain*> getMmaRootDimensions(
+    TensorView* tv,
+    MmaOp* mma,
+    MmaDimension dimension) {
+  // Build a fusion-level root domain map
+  //  so we can use the mma swizzles on non-immediate tensor operands, for
+  //  example loadstore staging ops.
+  ComputeAtRootDomainMap root_map;
+  root_map.build();
+
+  // FIXME:
+  // Several optimization is possible at this stage but assuming we don't have
+  //  a lot of mma ops in a fusion this could be lower priority.
+  // First it'd be nice not having to build root map every time this function
+  //  is called. That'd require some explicit boundary where we "lock" the
+  //  compute in the fusion so the root map stays valid.
+  // Second it'd reduce complexity of the below matching by an order if we have
+  //  something similar to "disjointSetOf" in idGraph, for just the root domains
+  //  at scheduler composing time.
```

Comment:
All these utilities are used in scheduling time. `RootDomainMap` should be completely defined since the compute is defined. But computeAt may still change after this stage so not sure if we'd want to be building an IdGraph at this stage. Also seems to be quite compute heavy to do in sync with scheduler primitives.